### PR TITLE
test(eject): observe the ejected app instead of grepping it

### LINKED ISSUE #7133: refactor: centralize polyglot output infrastructure (eject/bundle/nacompile artifact layer)
## Summary

Jac now has several features that turn `mod.gen.*` into traditional polyglot artifacts on disk: `jac eject` (FastAPI + Vite app), `jac bundle` (wheel/sdist/npm), `jac nacompile` (native/wasm binaries), serve-time client bundling, and the `.jir` cache. The compiler layer feeding these is well centralized — every backend (`PyastGenPass`, `EsastGenPass`, `NaIRGenPass`) is a real pass in the single schedule (`jac0core/compiler.jac` `get_py_code_gen()`) writing into the single `CodeGenTarget` container (`jac0core/codeinfo.jac`). The **artifact layer above it is not**: region routing, module classification, import rewriting, file layout, and app scaffolding are each implemented multiple times, in diverging ways, across `cli/commands/eject*`, `publish/`, `runtimelib/client/`, and `compiler/passes/native/`.

This is a tracking issue for consolidating that layer. Besides paying down duplication, this is deliberate groundwork for emitting additional languages/targets from Jac code later — new outputs should plug into one artifact pipeline, not add a seventh copy of it.

## Current duplication inventory

**Region (cl/sv/na) splitting exists in four mechanisms:**
1. `ModuleCodegenPass.iter_context_stmts` (`jac0core/passes/module_codegen_pass.jac:36`) — the intended central API; used by the native pass and capability checks.
2. `EsastGenPass` bypasses it with inline `code_context == CLIENT` checks scattered through `compiler/passes/ecmascript/impl/esast_gen_pass.impl.jac`.
3. Parse-time whole-file coercion `_coerce_{client,server,native}_module` (`jac0core/compiler.jac:305`), independently re-invoked by `na_ir_gen_pass.impl/core.impl.jac` and `annex_pass.impl.jac`.
4. JIR (de)serialization re-derives `code_context` again (`jac0core/impl/jir_passes.impl.jac`).

**Module classification diverges between eject and bundle:**
- Bundle uses the canonical `is_server_module()` / `is_annex()` from `ext_registry` (`publish/impl/builder.impl.jac:190`), which correctly excludes `.na.jac` / `.impl.jac`.
- Eject rolls its own heuristics: `_is_meaningful_python()` (scan for first non-import line, `cli/commands/impl/eject.impl.jac:355`) and `_js_exports_app()` (whitespace-strips generated JS and pattern-matches literals like `exportconstapp=`, `eject.impl.jac:328`).
- **Bug:** eject classifies files only as client-vs-not, so `.na.jac` modules are silently routed to the backend as *server* modules (or dropped) instead of being excluded/handled.

**Import rewriting implemented twice:** AST-based `py_imports.rewrite_imports` (wheel path) vs line-prefix `_rewrite_imports` (`cli/commands/eject_targets/backend.jac:184`).

**Source-file collection implemented three times:** `publish/impl/sources.impl.jac`, `publish/impl/npm_sources.impl.jac`, and eject's `_collect_jac_files`/`_dir_excluded`.

**Artifact layout/naming decided in ~6 independent writers** with no shared module: `.jir` cache (`jac0core/impl/compiler.impl.jac` + `jir.jac`), `nacompile` binary naming, serve-time client compiler (`.jac/client/compiled/`), eject's `backend/`+`frontend/` tree, wheel/npm builders, scale injector (`scale/injector/bundle.jac`). `es_to_js` is invoked from three separate non-test call sites.

**Large string-template emitters:** eject's `_emit_main_py` (`eject_targets/backend.jac:345` — ~300 lines of FastAPI auth/SSE/endpoint code as string concatenation, a shadow copy of serve behavior), plus hand-templated `package.json`/`vite.config.js`/`index.html`, the wheel's PEP 517 backend string globals, and `nacompile`'s LLVM entry-point IR text templates.

**Native target selection is ad hoc:** `JAC_NATIVE_TARGET` env var + name→triple mapping and per-OS if-ladders in `cli/commands/impl/nacompile.impl.jac` (target map ~line 155; linker/entry-point ladder ~lines 364–566); `if "wasm" in triple` guards inside codegen. No target-descriptor abstraction.

---

## Phase 1 — Eject correctness & convergence with publish

Small, high value; fixes a real misrouting bug.

- [ ] Replace `_is_meaningful_python()` with `is_server_module()`/`is_annex()` from `ext_registry` (same as the wheel builder).
- [ ] Explicitly handle `.na.jac` under eject: exclude with a warning (matching wheel behavior) as a first step; proper native artifact support tracked separately.
- [ ] Replace `_js_exports_app()` string-scraping with `mod.gen.client_manifest` (already produced by the compiler and consumed by the `.d.ts` generator in `publish/dts.jac`).
- [ ] Replace eject's `_rewrite_imports` with the AST-based `py_imports.rewrite_imports`.
- [ ] Extract one shared source-collection helper (package globs + excluded dirs) used by wheel, npm, and eject.
- [ ] Deduplicate `runtime_pkg_name` (`npm_sources.impl.jac` vs `runtime_npm.impl.jac`).
- [ ] Remove silent `except Exception { _ = 0; }` swallows in `eject.impl.jac` (frontend dir removal, sourcemap copy, runtime read) — warn instead.
- [ ] Tests: `.na.jac` fixture under eject; frontend detection driven by `client_manifest` (including formatted/edge-case JS that broke the old heuristic).

## Phase 2 — Single region-splitting API

- [ ] Refactor `EsastGenPass` onto `iter_context_stmts` / `shared_archetypes_for_context`; eliminate the inline `code_context == CLIENT` checks.
- [ ] Coerce codespace modules exactly once at parse; downstream consumers (`na_ir_gen_pass`, `annex_pass`) trust the stamped `code_context` instead of re-invoking `_coerce_*_module`.
- [ ] JIR round-trips `code_context` instead of re-deriving it on load.
- [ ] Acceptance: a change to region semantics requires touching one module; byte-identical `gen.py`/`gen.js`/`gen.llvm_ir` on the existing test corpus before/after.

## Phase 3 — Thin ejected `main.py`

- [ ] Extract FastAPI app construction (auth, SSE, walker/function endpoint registration, static mount) into a runtime app-factory that shares its core with `scale/server/serve.jac`.
- [ ] `_emit_main_py` shrinks to a thin generated entrypoint importing the vendored factory; no endpoint/auth/SSE behavior encoded as string literals.
- [ ] Fix the cross-module walker-inheritance field drop (`_collect_walker_fields`) by moving to runtime introspection instead of source-time AST scraping.
- [ ] Acceptance: ejected-app endpoint behavior is exercised by the same code paths (and ideally tests) as `jac serve`.

## Phase 4 — Shared artifact-layout module

- [ ] One module owning output path/naming/layout decisions, consumed by the `.jir` writer, `nacompile`, the serve-time client compiler, eject, wheel/npm builders, and the scale injector.
- [ ] A single "emit module → target artifacts" entry point so CLI/serve/eject/publish stop each re-orchestrating compile-and-write.
- [ ] Consolidate `es_to_js` call sites behind it.

## Phase 5 — Native target-descriptor registry

- [ ] Introduce a target-descriptor object (triple, data layout, entry-point injector, linker class, runtime-shim policy, codegen quirks) and a registry keyed by target name.
- [ ] Replace the `JAC_NATIVE_TARGET` name→triple map and per-OS if-ladders in `nacompile.impl.jac` with registry lookups; `if "wasm" in triple` guards in `na_ir_gen_pass` become descriptor queries.
- [ ] Acceptance: adding a hypothetical new target touches the registry + a linker, not the CLI driver.

## Phase 6 — Test-coverage gaps

- [ ] `ViteBundler` / production client-bundle path.
- [ ] `npm-runtime` and framework-variant runtime tarballs.
- [ ] `bundle --target npm` / `all` at the CLI layer (currently only `NpmBuilder` unit tests).
- [ ] `runtime_vendor.jac` standalone unit tests (currently covered only transitively via eject/publish).

## Non-goals

- New output languages (e.g. C source backend) and new native targets — separate roadmap; this issue is the groundwork it plugs into.
- Changing artifact *formats* (wheel/npm/wasm layouts stay as-is; only where the logic lives changes).

## Suggested sequencing

Phases 1 and 5 are independent and can start immediately. Phase 2 should land before Phase 4 (the artifact layer wants one region API underneath it). Phase 3 is independent of 2/4 but benefits from Phase 1's classification cleanup. Phase 6 accumulates alongside each phase rather than at the end.

### PR DESCRIPTION
**Pull request by kashmithnisakya**
Groundwork for #7133 (PR 1 of the plan in [this comment](https://github.com/jaseci-labs/jac/issues/7133#issuecomment-4959480550)).

## Why

All 46 tests in `test_eject.jac` are text assertions on the generated `main.py`:

```jac
assert '@app.post("/walker/Greet")' in main_py;
```

Not one of them boots the app or issues a request. So the ~400-line FastAPI server that `_emit_main_py` writes out as string literals has **zero behavioral coverage**, and there is no CLI-level test for `--as source` at all. The #7133 fixes all touch that emitter; right now nothing would catch a regression in it.

## What

Two nets, because they catch different things.

**1. `_EjectedApp`, an in-process harness.** Ejects into a `TemporaryDirectory()`, imports the generated backend, and drives it through fastapi's `TestClient`. Six tests:

- the app boots and mounts the routes it declares
- register then login mints a token, and a wrong password does not
- a `:pub` walker runs unauthenticated and reports its result
- a token-gated walker 401s without a token, and runs with one
- a `:pub` function runs and returns its value
- booting one app does not leak its modules into the next

That last one is not paranoia. An ejected backend uses generic module names (`main`, `server`), the suite runs in one process, and a leak would silently serve the previous test's app to the next test. Teardown purges the backend's modules from `sys.modules` and drops its dir from `sys.path`, and the test asserts it.

**2. `ejected-app-smoke`, a nightly CI job.** Ejects the `todo_app` example, `pip install`s it into a **clean venv**, boots it under uvicorn, and drives it over real HTTP: register, login, `create_todo`, then `read_todos` returns the persisted node.

The clean venv is the whole point. It is the only thing that exercises the runtime **vendored into `backend/`**, because the in-process harness binds to the live `jaclang` already in `sys.modules` and never touches the vendored copy. `runtime_vendor.jac` decides what ships inside every artifact and currently has zero direct tests.

Worth stating plainly: running the command is not itself a test. `jac build --as source` exits 0 on output that cannot boot, so the job has to boot it and hit it or it proves nothing.

## Does the net have holes the right size?

Yes, verified against a walker that raises:

```
>>> STATUS: 200
>>> BODY:   {"data": {"error": "kaboom", "traceback": "Traceback (most recent call last)..."}}
```

The ejected app serves a crashed walker as `200 OK` with the Python traceback in the response body, where `jac start` returns a 500 `TransportResponse.fail`. That test lands with its fix, not in this PR.

## Also found

`jac/examples/littleX` ejects to a backend with **0 walkers and 0 functions**, despite `social_graph.jac` declaring 20 walkers. Its `main.jac` is a client-only `cl { }` entry, and the walkers are reached by `sv import` from the client module, which eject's walker discovery does not follow. So the flagship fullstack example currently ejects to a non-functional backend. That is why this PR's CI job uses `todo_app` (3 walkers, server-reachable) instead. Raising it separately on #7133.

## Notes

- Tests only, no product code changes.
- 52 passed (46 existing + 6 new). No new typecheck errors (50 on `main`, 50 here).

### CODE REVIEW

File: jac/tests/project/test_eject.jac [hunk 1]
```diff
@@ -71,6 +71,75 @@ def _has_jaclang_import(py_text: str) -> bool {
...
+    has src: Path,
+        backend: str = "",
+        preexisting: set[str] = set();
+
+    def __enter__ -> any {
+        import importlib;
+        import from fastapi.testclient { TestClient }
+        self.backend = str((self.src / "ejected" / "backend").resolve());
+        self.preexisting = set(sys.modules.keys());
+        sys.path.insert(0, self.backend);
```
[the remark this answers is not in the document]
Comment (kashmithnisakya) [reply jac/tests/project/test_eject.jac]: good catch, this was real. confirmed it by seeding sys.modules with a decoy `main` before booting - the harness bound to it and the test failed.

__enter__ now evicts colliding top-level names from the ejected backend before importing, except `jaclang` which has to stay on the live runtime rather than rebinding to the vendored copy. preexisting is captured after the eviction so the name still gets purged on exit. added the decoy test as a regression, it fails without the fix.

File: jac/tests/project/test_eject_boot.jac [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            if entry.suffix == ".py" {
+                name = entry.stem;
+            } elif entry.is_dir() {
+                name = entry.name;
+            } else {
+                continue;
+            }
+            if name != "jaclang" {
+                sys.modules.pop(name, None);
+            }
```
[the remark this answers is not in the document]
Comment (kashmithnisakya) [reply jac/tests/project/test_eject_boot.jac]: right again, and it was reachable. reproduced it with a package-shaped fixture (eject emits lib/__init__.py + lib/helpers.py, so sys.modules keys the child as `lib.helpers`): with a decoy cached the walker returned "STALE" instead of "ADA".

eviction now drops the dotted children too, not just the top-level name, still exempting `jaclang`. added it as a regression test - it fails without the fix.

**Comment by marsninja** [comment]
Closing: no activity in 10+ days. Reopen if still relevant.
